src/utils.ts

```diff
@@ -898,97 +898,224 @@ export function deepCopy<T>(value: T): T {
+    return 'List' as const;
+  }
+
+  constructor() {
+    this.count = 0;
+
+    const sentinel = {
+      value: null as unknown as T,
+      next: null as unknown as ListNode<T>,
+      prev: null as unknown as ListNode<T>
+    };
+    sentinel.next = sentinel as unknown as ListNode<T>;
+    sentinel.prev = sentinel as unknown as ListNode<T>;
+
+    this.head = sentinel;
+  }
+
+  toArray() {
+    return Array.from(this);
+  }
+
+  toString() {
+    return `head <=> ${this.toArray().join(' <=> ')} <=> head`;
+  }
+
+  *[Symbol.iterator](): Generator<T, void, void> {
+    for (const node of this.nodes()) {
+      yield node.value;
+    }
+  }
+
+  private *nodes() {
+    let ptr = this.head.next;
+    while (ptr !== this.head) {
+      // Save next before yielding so that we make removing within iteration safe
+      const { next } = ptr;
+      yield ptr;
+      ptr = next;
+    }
+  }
+
+  /** Insert at end of list */
+  push(value: T) {
+    this.count += 1;
+    const newNode: ListNode<T> = {
+      next: this.head as unknown as ListNode<T>,
+      prev: this.head.prev,
+      value
+    };
+    this.head.prev.next = newNode;
+    this.head.prev = newNode;
+  }
+
+  /** Inserts every item inside an iterable instead of the iterable itself */
+  pushMany(iterable: Iterable<T>) {
+    for (const value of iterable) {
+      this.push(value);
+    }
+  }
+
+  /** Insert at front of list */
+  unshift(value: T) {
+    this.count += 1;
+    const newNode: ListNode<T> = {
+      next: this.head.next,
+      prev: this.head as unknown as ListNode<T>,
+      value
+    };
+    this.head.next.prev = newNode;
+    this.head.next = newNode;
+  }
+
+  private remove(node?: ListNode<T> | null) {
+    if (node == null || this.length === 0) {
+      return { next: null, prev: null, value: null };
+    }
+
+    this.count -= 1;
+
+    const prevNode = node.prev;
+    const nextNode = node.next;
+    prevNode.next = nextNode;
+    nextNode.prev = prevNode;
+
+    // Null out the pointers so that a double remove does not break things
+    node.next = null as any;
+    node.prev = null as any;
+    return node;
+  }
+
+  /** Removes the first node at the front of the list */
+  shift() {
+    return this.remove(this.head.next).value;
+  }
+
+  /** Removes the last node at the end of the list */
+  pop() {
+    return this.remove(this.head.prev).value;
+  }
+
+  /** Iterates through the list and removes nodes where filter returns true */
+  prune(filter: (value: T) => boolean) {
+    for (const node of this.nodes()) {
+      if (filter(node.value)) {
+        this.remove(node);
+      }
+    }
+  }
+
+  clear() {
+    this.count = 0;
+    this.head.next = this.head;
+    this.head.prev = this.head;
+  }
+
+  first(): T | null {
+    // If the list is empty, value will be the sentinel's null
+    return this.head.next.value;
+  }
+
+  last(): T | null {
+    // If the list is empty, value will be the sentinel's null
+    return this.head.prev.value;
```

Comment:
personal preference: instead of relying on the comment to explain the clever code, we could make the code self-documenting and the comment unnecessary if we wrote it like this:

```typescript
return this.length === 0 ? null : this.head.prev.value;
```

same comment applies to `first()`.

Reply:
I don't think adding code helps document something, I can further explain what I mean in the comment but adding code is adding steps for no functional gain, we have a null saved at `head.value` so we might as well use it

Reply:
This is a matter of taste, I don't think it is particularly clever to use a variable we have access to and we test for. This is currently branchless logic, while the developer has to read a bit to catch the case they may not have considered (and the comment intends to help them with that) it is less runtime steps to accomplish the goal of the method

Reply:
With the TS suggestions Daria has made the type of head is now correctly null so that lends more documentation to this behavior